### PR TITLE
Add $view->registerScript()

### PR DESCRIPTION
**Pull request by khalwat**
* Add $view->registerScript()
* Add render <title> tag if $view->title is set

Resolves this: yiisoft/yii2#11018

It's resolved in such a way that even if they do decide to implement some method of adding non application/javascript script tags, it won't conflict.

### CODE REVIEW

File: src/web/View.php [hunk 1]
```diff
@@ -643,6 +649,48 @@ public function clearJsBuffer(bool $scriptTag = true)
...
+    }
+
+    /**
+     * @inheritdoc
+     */
+    protected function renderHeadHtml()
+    {
+        $lines = [];
+        if (!empty($this->title)) {
+            $lines[] = '<title>' . Html::encode($this->title) . '</title>' . "\n";
```
Comment (brandonkelly) [inline src/web/View.php]: This doesn't seem related to the rest of the PR, unless I’m missing something?

File: src/web/View.php [hunk 2]
```diff
@@ -643,6 +649,48 @@ public function clearJsBuffer(bool $scriptTag = true)
...
 
     /**
+     * Registers a generic <script> code block.
+     * @param string $script the generic <script> code block to be registered
+     *
+     * @param string $key the key that identifies the generic <script> code block. If null, it will use
+     * $script as the key. If two generic <script> code blocks are registered with the same key, the latter
+     * will overwrite the former.
+     */
+    public function registerScript($script, $options = [], $key = null)
```
Comment (brandonkelly) [inline src/web/View.php]: There should be a `$position` argument here just like `registerJs()`, but only which only supports `POS_HEAD`, `POS_BEGIN`, and `POS_END`.

File: src/web/View.php [hunk 3]
```diff
@@ -643,6 +649,48 @@ public function clearJsBuffer(bool $scriptTag = true)
...
+    }
+
+    /**
+     * @inheritdoc
+     */
+    protected function renderBodyEndHtml($ajaxMode)
+    {
+        $lines = [];
+        if (!empty($this->_scripts)) {
+            $lines[] = implode("\n", $this->_scripts);
```
Comment (brandonkelly) [inline src/web/View.php]: This would result in

```php
$lines = [
    [
        '<script ... />',
        // ...
    ]
];
```

rather than

```php
$lines = [
    '<script ... />',
    // ...
];
```

File: src/web/View.php (hunk 1 above)
Comment (khalwat) [inline src/web/View.php]: It's something else that I'm hoping can be added. I can do a separate PR if you like; it emulates what Yii2 does in their default template `yiisoft/yii2-debug/views/layouts/main.php`

If `$title` is not set (as is the case by default), then nothing happens. If it is set, then it renders the `<title>` tag based on `$view->title`

File: src/web/View.php (hunk 3 above)
Comment (khalwat) [inline src/web/View.php]: Well, the array ends up looking like this:

`Craft::dd($lines);`

```
[
    0 => '<script type=\"application/ld+json\">{
    \"@context\": \"http://schema.org\",
    \"@type\": \"Article\",
    \"name\": \"Some Blog\",
    \"url\": \"https://nystudio107.com/blog\"
}</script>'
]```

File: src/web/View.php (hunk 2 above)
Comment (khalwat) [inline src/web/View.php]: yep, okay. Requires extending a couple more methods, but I'd agree, it's better this way.

File: src/web/View.php (hunk 3 above)
Comment (khalwat) [inline src/web/View.php]: ah, I see. Nevermind.

Comment (khalwat) [inline src/web/View.php]: Okay so for whatever reason, this is the code pattern that Yii2 uses, so I just followed it. They end up doing an `implode("\n", $lines)` at the end, which results in it being formatted properly.

Example from `yii\web\view`:

```
    protected function renderHeadHtml()
    {
        $lines = [];
        if (!empty($this->metaTags)) {
            $lines[] = implode("\n", $this->metaTags);
        }

        if (!empty($this->linkTags)) {
            $lines[] = implode("\n", $this->linkTags);
        }
        if (!empty($this->cssFiles)) {
            $lines[] = implode("\n", $this->cssFiles);
        }
        if (!empty($this->css)) {
            $lines[] = implode("\n", $this->css);
        }
        if (!empty($this->jsFiles[self::POS_HEAD])) {
            $lines[] = implode("\n", $this->jsFiles[self::POS_HEAD]);
        }
        if (!empty($this->js[self::POS_HEAD])) {
            $lines[] = Html::script(implode("\n", $this->js[self::POS_HEAD]), ['type' => 'text/javascript']);
        }

        return empty($lines) ? '' : implode("\n", $lines);
    }
```

**Comment by khalwat** [comment]
Requested changes made

File: src/web/View.php (hunk 1 above)
Comment (brandonkelly) [inline src/web/View.php]: I'd rather see this as a separate PR; it's confusing in this context.

File: src/web/View.php (hunk 3 above)
Comment (brandonkelly) [inline src/web/View.php]: OK gotcha. It's a bit confusing at first glance, but I guess consistency with Yii is a good thing.

**Comment by brandonkelly** [comment]
All looks great, and I'm not opposed to the `<title>` thing, but that belongs in a separate PR.

**Comment by khalwat** [comment]
Removed the `<title>` rendering (it's just 2 lines of code). Will submit via a separate PR.

**Comment by brandonkelly** [comment]
Thank you sir!
